### PR TITLE
Actions rule config UI

### PR DESCRIPTION
**Pull request by eriksendc**
Summary
---------

This PR introduces a first draft of the Action Rules (in Settings). 

This PR is organized as a "stacked diff". It's easiest to review per commit:

1. Added the concept of a route at the tab level so that page refreshes don't go back to the default tab.
2. Added the Action Rules tab without any real functionality (just getting the main component in place).
3. Added the main functionality: a table with the rules configured so far; the ability to add a rule; editing and deleting a rule is mocked (incomplete functionality).
4. Added validation: Name required and must be unique; Labeled As required; Action required.
5. (skip - just fixing formatting)
6. Added delete functionality.
7. Added edit functionality.

Test Plan
---------

Run the web app, navigate to Settings --> Action Rules then exercise the UI. You should see what's captured in the attached video.

https://user-images.githubusercontent.com/8974354/116909103-b79eef00-ac11-11eb-933a-28e7bab014f0.mp4

### CODE REVIEW

File: hasher-matcher-actioner/webapp/src/App.jsx
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/App.jsx:54]: Yay! Love this

File: hasher-matcher-actioner/webapp/src/Settings.jsx
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/Settings.jsx]: How difficult would it be to read this tab list from somewhere because I already know its going to change (Mengyang has a ThreatExchange Tab coming)

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:263]: is `{' '}` on purpose?

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:411]: .

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (jeberl) [inline hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:62]: So we cant have a rule with only must_not_have_labels?

**Review by BarrettOlson — COMMENTED** [review]
Reviewed the first few commits and didn't notice anything to RC about though I added some nits inline.

I think it would be helpful to break up the ActionRuleSettingsTab.jsx file and it has become pretty complex.  
Also the webapp directory structure is maybe a topic for discussion.

**Review by jeberl — APPROVED** [review]
This looks good to me. Some comments in line you might want to think about

File: hasher-matcher-actioner/webapp/src/Settings.jsx
Comment (schatten) [inline hasher-matcher-actioner/webapp/src/Settings.jsx]: Wouldn't [`defaultActiveKey`](https://react-bootstrap.github.io/components/tabs/#tabs-api) in <Tabs> solve this issue?

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (schatten) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:1]: More context [here](https://github.com/facebook/ThreatExchange/pull/546#pullrequestreview-651286235), @eriksendc . Apologies for not having updated README with this yet. 

Summary:
-  webapp/src, there are two directories, `components/` and `pages/`.
- All top-level pages should go into `pages/`, this includes settings.
   - A top-level page is one that calls APIs eg. settings, matches, dashboard
- Components required by those pages that are either large-enough or re-useable can be moved to `components/`

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (schatten) [inline hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:121]: Not following this. Is it a common pattern now to have refs as 'state'?

I have seen `useRef()` in the past, and simply `const x = React.createRef()`, but this is new.

Ofcourse, my knowledge of react may have already been outdated. 😓

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (schatten) [inline hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:231]: This is highly unorthodox. [Controlled inputs](https://reactjs.org/docs/forms.html#controlled-components) in react would update their parent's state with their current values when things change. 

Using a ref to obtain current state information defeats the purpose of state.

Consider using controlled components.

Comment (schatten) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:231]: Retaining refs to sibling components is especially dangerous. React internals 'choose' when to update state. It may use requestAnimationFrame to debounce frequent state updates. So your ref may be out dated if it is referenced any where except the component that defines it.

If there is one thing you take from all of this, avoid refs for state. Only use if it you need to do native events like focus().

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (schatten) [inline hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx]: nit: The onClick handler is getting too large to read easily. Consider moving it to its own function and simply calling that function in onClick?

**Review by schatten — CHANGES_REQUESTED** [review]
Overall, love the UI and how it helps users "course-correct" with the validations.

Have serious concerns about the usage of refs instead of [lifting-state-up](https://reactjs.org/docs/lifting-state-up.html).

Putting it back into your queue.

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:263]: Yes. Prettier does that as one of its standards.

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:411]: .

**Review by schatten — APPROVED** [review]
@eriksendc My previous strong opinion about not using refs has been addressed. Happy to unblock.

As you have already noted offline, there are some outstanding comments from @BarrettOlson and myself. I'd be okay if you want to deal with them separately OR in this PR as additional commits. Your call.

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:1]: Resolved in subsequent pr

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:46]: great! done in subsequent pr! :)

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:326]: .

File: hasher-matcher-actioner/webapp/src/Settings.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/Settings.jsx]: good call! done in a subsequent PR.

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx]: That's right. I made that decision for now... I couldn't envision blindly accepting content to action on without some kind of "positive" classification criteria. Easy to revisit later, if we find a driving use case.

File: hasher-matcher-actioner/webapp/src/Settings.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/Settings.jsx]: as discussed this was part of the implementation to make tabs have routes.

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx]: no longer applicable. reworked to not use refs.

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx]: refactored to not use refs.

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx]: The exact issue went away when I did a subsequent refactor, but this comment was still helpful... look for defaultActionRule being defined in a subsequent PR.

File: hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx
Comment (eriksendc) [reply hasher-matcher-actioner/webapp/src/settings/ActionRuleSettingsTab.jsx:5]: done in PR 562.

**Pull request merged**
